### DISCUSSION TITLE
Add .deb / APT / PPA support to Packages

### DISCUSSION THREAD

**Discussion by asasine** [community/community#56445] [Packages]
### Select Topic Area

Product Feedback

### Summary
I'd like to be able to create Debian binary packages and upload them to GitHub Packages.

### Intended Outcome
Users can upload Debian binary packages (*.deb) to GitHub Packages and add a repo's package feed as a custom APT source.

### How could it work?
1. The user uploads a Debian binary package to GitHub Packages, possible through a step in GitHub Actions.
2. The user adds the GitHub Packages registry to their APT sources on their Debian machine:
    ```
    # /etc/apt/sources.list.d/my-github-packages.list
    deb https://apt.pkg.github.com/<username>/<repo> stable main
    ```

4. The user installs packages from this new source using conventional commands:
    ```bash
    sudo apt update
    sudo apt install my-github-package
    ```

### Bonus features
Authentication can be done inline to the APT source with a PAT.

```
# /etc/apt/sources.list.d/my-github-packages.list
deb https://<username>:<PAT>@apt.pkg.github.com/<username>/<repo> stable main
```

Instructions could point out the use of [apt_auth.conf(5)](https://manpages.debian.org/testing/apt/apt_auth.conf.5.en.html) to separate secrets from the APT source lists.

Authentication could also be done with SSH-based sources

```
# /etc/apt/sources.list.d/my-github-packages.list
deb ssh://<username>@apt.pkg.github.com/<username>/<repo> stable main
```

**Comment by metux** [comment]
I'm looking for the same, too.

**Comment by mehdiben7** [comment]
This would be awesome ! If that can help Github developpers, I know that Gitea already supports it (and Gitea is open source). Possibily also Gitlab. I would love to be able to do the same on Github

**Comment by danielsiegl** [comment]
I also want this

**Comment by MickaelOnTheWave** [comment]
Me too

**Comment by asmuiahmad** [comment]
Are you able to add repositories via the Software Source Menu (Ubuntu Software Center -> Edit -> software sources -> other software -> add)? If were can, then you would perform "sudo apt-get update" to update your repo information. Then after that, just type in sudo apt-get install 'package name'(be as specific as possible as apt-get is very finicky)

**Reply by asasine** [reply]
GitHub Packages does not have the ability to be an APT repo source. Even if you have permissions on your machine to update the APT repo sources, you cannot point it towards GitHub Packages.

**Comment by nnarain** [comment]
I'll throw my support in here too. 

I'm working on a little robot project. I deploy the software using docker pulled from GitHub. I use a debian for udev rules and systemd services that manage the docker stack. I use a script to pull the latest debian from GitHub and update.

So it's all on GitHub already. The missing convenience is using apt to pull the updates in.

**Comment by asasine** [comment]
Not an answer, just a workaround for those interested: I created [`apt-transport-github`](https://github.com/asasine/apt-transport-github), which uses APT's modular transport functionality to fetch Debian binary packages from attached Assets in GitHub Releases of any repo whenever you run `apt-get update`, making them available for installation. By no means is it a complete replacement for a full APT repository, but it does a decent job of filling this gap in GitHub Packages. My apologies if this looks like self-promotion, just trying to help people out. If it looks useful but doesn't cover your scenario, create an issue or discussion.
